src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,364 @@
+            {
+                ThrowHelper.ThrowBadImageFormatException(InvalidUnsafeAccessorUsage);
+            }
+
+            TypeDesc declTypeDesc;
+            TypeDesc maybeTypeDesc;
+            foreach (MethodDesc md in targetType.GetMethods())
+            {
+                // Check the target and current method match static/instance state.
+                if (context.IsTargetStatic != md.Signature.IsStatic)
+                {
+                    continue;
+                }
+
+                // Check for matching name
+                if (!md.Name.Equals(name))
+                {
+                    continue;
+                }
+
+                // Validate calling convention.
+                if ((MethodSignatureFlags.UnmanagedCallingConventionMask & md.Signature.Flags)
+                    != (MethodSignatureFlags.UnmanagedCallingConventionMask & context.Declaration.Signature.Flags))
+                {
+                    continue;
+                }
+
+                // Validate the return type and prepare for validating
+                // the current signature's argument list.
+                int sigCount = context.Declaration.Signature.Length;
+                if (context.Kind == UnsafeAccessorKind.Constructor)
+                {
+                    if (!md.Signature.ReturnType.IsVoid)
+                    {
+                        continue;
+                    }
+                }
+                else
+                {
+                    declTypeDesc = context.Declaration.Signature.ReturnType;
+                    maybeTypeDesc = md.Signature.ReturnType;
+                    if (declTypeDesc != maybeTypeDesc)
+                    {
+                        continue;
+                    }
+
+                    // Non-constructor accessors skip the first argument
+                    // when validating the target argument list.
+                    sigCount--;
+                }
+
+                // Validate argument count matches.
+                if (sigCount != md.Signature.Length)
+                {
+                    continue;
+                }
+
+                // Validate arguments match - reverse order
+                for (; sigCount > 0; --sigCount)
+                {
+                    declTypeDesc = context.Declaration.Signature[sigCount];
+                    maybeTypeDesc = md.Signature[sigCount];
+                    if (declTypeDesc != maybeTypeDesc)
+                    {
+                        break;
+                    }
+                }
+
+                // If we validated all arguments, we have a match.
+                if (sigCount != 0)
+                {
+                    continue;
+                }
+
+                context.TargetMethod = md;
+                return true;
+            }
+            return false;
+        }
+
+        private static bool TrySetTargetMethodCtor(ref GenerationContext context)
+        {
```

Comment:
This can use error handling for the typedesc cases too. For example, this:
```
using System.Runtime.CompilerServices;

class Test
{
    public Test()
    {
    }

    [UnsafeAccessor(UnsafeAccessorKind.Constructor)]
    extern static ref Test CtorAccessor();

    static void Main()
    {
        CtorAccessor();
    }
}
```
fails with:
```
Assert failure(PID 34828 [0x0000880c], Thread: 14728 [0x3988]): Consistency check failed: AV in clr at this callstack:
------
CORECLR! `anonymous namespace'::TrySetTargetMethodCtor + 0xF8 (0x00007fff`013a71a8)
CORECLR! MethodDesc::TryGenerateUnsafeAccessor + 0x54E (0x00007fff`013a65ee)
CORECLR! getMethodInfoHelper + 0x14B7 (0x00007fff`012f3fc7)
CORECLR! UnsafeJitFunction + 0x353 (0x00007fff`012cfe43)
```